### PR TITLE
ECL-628: Complete service refactor

### PR DESCRIPTION
**Pull request by liandra-giacomine**

### CODE REVIEW

File: app/uk/gov/hmrc/economiccrimelevyregistration/connectors/EmailConnector.scala [hunk 1]
```diff
@@ -41,42 +51,56 @@ class EmailConnector @Inject() (appConfig: AppConfig, httpClient: HttpClient)(im
...
-        RegistrationSubmittedEmailRequest(
-          to = Seq(to),
-          templateId = entityType match {
-            case Some(value) if EntityType.isOther(value) =>
-              OtherEntityTemplateId
-            case _                                        =>
-              NormalEntityTemplateId
-          },
-          parameters = registrationSubmittedEmailParameters
+    retryFor[Unit]("HMRC email - Send registration")(retryCondition) { //TODO - do we need a limit for the amount of retries?
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/economiccrimelevyregistration/connectors/EmailConnector.scala:54]: Please can you resolve or remove the TODO.

File: app/uk/gov/hmrc/economiccrimelevyregistration/connectors/IncorporatedEntityIdentificationFrontendConnector.scala [hunk 2]
```diff
@@ -36,39 +40,62 @@ trait IncorporatedEntityIdentificationFrontendConnector {
...
     val serviceNameLabels = ServiceNameLabels()
 
-    val url: String = incorporatedEntityType match {
-      case UkLimitedCompany | UnlimitedCompany => s"$apiUrl/limited-company-journey"
-      case RegisteredSociety                   => s"$apiUrl/registered-society-journey"
-      case e                                   => throw new IllegalArgumentException(s"$e is not a valid incorporated entity type")
+    val url: URL = incorporatedEntityType match {
+      case UkLimitedCompany | UnlimitedCompany => url"$apiUrl/limited-company-journey"
+      case RegisteredSociety                   => url"$apiUrl/registered-society-journey"
+      case _                                   => url""
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/economiccrimelevyregistration/connectors/IncorporatedEntityIdentificationFrontendConnector.scala:64]: Should this not be erroring?

File: app/uk/gov/hmrc/economiccrimelevyregistration/connectors/PartnershipIdentificationFrontendConnector.scala [hunk 3]
```diff
@@ -38,42 +42,61 @@ trait PartnershipIdentificationFrontendConnector {
...
-      case ScottishLimitedPartnership  => s"$apiUrl/scottish-limited-partnership-journey"
-      case LimitedLiabilityPartnership => s"$apiUrl/limited-liability-partnership-journey"
-      case e                           => throw new IllegalArgumentException(s"$e is not a valid partnership type")
+    val url: URL = partnershipType match {
+      case GeneralPartnership          => url"$apiUrl/general-partnership-journey"
+      case ScottishPartnership         => url"$apiUrl/scottish-partnership-journey"
+      case LimitedPartnership          => url"$apiUrl/limited-partnership-journey"
+      case ScottishLimitedPartnership  => url"$apiUrl/scottish-limited-partnership-journey"
+      case LimitedLiabilityPartnership => url"$apiUrl/limited-liability-partnership-journey"
+      case _                           => url""
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/economiccrimelevyregistration/connectors/PartnershipIdentificationFrontendConnector.scala:70]: Should this not be erroring?

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/FirstContactEmailController.scala [hunk 4]
```diff
@@ -37,26 +38,37 @@ class FirstContactEmailController @Inject() (
...
-        firstContactName(request),
-        mode,
-        request.registration.registrationType,
-        request.eclRegistrationReference
-      )
+    (for {
+      firstContactName <- request.firstContactNameOrError
+    } yield firstContactName).fold(
+      _ =>
+        Redirect(
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/FirstContactEmailController.scala:59]: Should this be routeError?

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/FirstContactNumberController.scala [hunk 5]
```diff
@@ -36,25 +38,37 @@ class FirstContactNumberController @Inject() (
...
-        request.registration.registrationType,
-        request.eclRegistrationReference
-      )
+    (for {
+      firstContactName <- request.firstContactNameOrError
+    } yield firstContactName).fold(
+      _ =>
+        Redirect(
+          uk.gov.hmrc.economiccrimelevyregistration.controllers.routes.NotableErrorController
+            .answersAreInvalid() //TODO - routeError instead
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/FirstContactNumberController.scala:60]: Please action or remove this comment.

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/FirstContactNumberController.scala [hunk 6]
```diff
@@ -63,28 +77,45 @@ class FirstContactNumberController @Inject() (
...
-            )
-          ),
+          (for {
+            firstContactName <- request.firstContactNameOrError
+          } yield firstContactName)
+            .fold(
+              _ =>
+                Future.successful(
+                  Redirect(
+                    uk.gov.hmrc.economiccrimelevyregistration.controllers.routes.NotableErrorController
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/FirstContactNumberController.scala:87]: Should this be routeError?

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/FirstContactRoleController.scala [hunk 7]
```diff
@@ -36,25 +38,36 @@ class FirstContactRoleController @Inject() (
...
-        firstContactName(request),
-        mode,
-        request.registration.registrationType,
-        request.eclRegistrationReference
-      )
+    (for {
+      firstContactName <- request.firstContactNameOrError
+    } yield firstContactName).fold(
+      _ =>
+        Redirect(
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/FirstContactRoleController.scala:58]: Should this be routeError?

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/FirstContactRoleController.scala [hunk 8]
```diff
@@ -63,26 +76,41 @@ class FirstContactRoleController @Inject() (
...
-              )
-            )
-          ),
+          (for {
+            firstContactName <- request.firstContactNameOrError
+          } yield firstContactName)
+            .fold(
+              _ =>
+                Future.successful(
+                  Redirect(
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/FirstContactRoleController.scala:85]: Should this be routeError?

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/SecondContactEmailController.scala [hunk 9]
```diff
@@ -36,25 +38,37 @@ class SecondContactEmailController @Inject() (
...
-        secondContactName(request),
-        mode,
-        request.registration.registrationType,
-        request.eclRegistrationReference
-      )
+    (for {
+      secondContactName <- request.secondContactNameOrError
+    } yield secondContactName).fold(
+      _ =>
+        Redirect(
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/SecondContactEmailController.scala:59]: Should this be routeError?

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/SecondContactEmailController.scala [hunk 10]
```diff
@@ -63,28 +77,48 @@ class SecondContactEmailController @Inject() (
...
-              )
-            )
-          ),
+          (for {
+            secondContactName <- request.secondContactNameOrError
+          } yield secondContactName)
+            .fold(
+              _ =>
+                Future.successful(
+                  Redirect(
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/SecondContactEmailController.scala:86]: Should this be routeError?

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/SecondContactNumberController.scala [hunk 11]
```diff
@@ -36,25 +38,36 @@ class SecondContactNumberController @Inject() (
...
-        secondContactName(request),
-        mode,
-        request.registration.registrationType,
-        request.eclRegistrationReference
-      )
+    (for {
+      secondContactName <- request.secondContactNameOrError
+    } yield secondContactName).fold(
+      _ =>
+        Redirect(
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/SecondContactNumberController.scala:58]: Should this be routeError?

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/SecondContactNumberController.scala [hunk 12]
```diff
@@ -63,28 +76,43 @@ class SecondContactNumberController @Inject() (
...
-              )
-            )
-          ),
+          (for {
+            secondContactName <- request.secondContactNameOrError
+          } yield secondContactName)
+            .fold(
+              _ =>
+                Future.successful(
+                  Redirect(
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/economiccrimelevyregistration/controllers/contacts/SecondContactNumberController.scala:85]: Should this be routeError?

File: app/uk/gov/hmrc/economiccrimelevyregistration/models/Registration.scala [hunk 13]
```diff
@@ -66,7 +66,7 @@ final case class Registration(
           region = address.region.map(_.trim),
           postCode = address.postal_code.map(_.trim),
           poBox = address.po_box.map(_.trim),
-          countryCode = "GB"
+          countryCode = "GB" //TODO - why is this hard coded?
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/economiccrimelevyregistration/models/Registration.scala:69]: Please action or remove comment.

File: app/uk/gov/hmrc/economiccrimelevyregistration/services/EclCalculatorService.scala [hunk 14]
```diff
@@ -16,44 +16,57 @@
...
 @Singleton
 class EclCalculatorService @Inject() (
   eclCalculatorConnector: EclCalculatorConnector
-)(implicit
-  ec: ExecutionContext
-) {
-  def checkIfRevenueMeetsThreshold(registration: Registration)(implicit hc: HeaderCarrier): Future[Option[Boolean]] =
+)(implicit ec: ExecutionContext) {
+
+//  TODO: separate calculateLiability logic from this method
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/economiccrimelevyregistration/services/EclCalculatorService.scala:35]: Please action or remove this comment.

**Review by TheDanielDoyle — CHANGES_REQUESTED** [review]
Well done. Only minor comments.

File: app/uk/gov/hmrc/economiccrimelevyregistration/connectors/IncorporatedEntityIdentificationFrontendConnector.scala (hunk 2 above)
Comment (liandra-giacomine) [reply app/uk/gov/hmrc/economiccrimelevyregistration/connectors/IncorporatedEntityIdentificationFrontendConnector.scala:64]: This gets called fro the service which checks whether it is one of those 3 entity types, so I removed the underscore case as it's not needed

Comment (TheDanielDoyle) [reply app/uk/gov/hmrc/economiccrimelevyregistration/connectors/IncorporatedEntityIdentificationFrontendConnector.scala:64]: Does removing the underscore case produce a warning though? If so, we need to handle that case somehow.

File: app/uk/gov/hmrc/economiccrimelevyregistration/connectors/PartnershipIdentificationFrontendConnector.scala (hunk 3 above)
Comment (liandra-giacomine) [reply app/uk/gov/hmrc/economiccrimelevyregistration/connectors/PartnershipIdentificationFrontendConnector.scala:70]: Same as above here

**Pull request merged**